### ISSUE TITLE
Esperanto stemmer

### ISSUE THREAD

**Issue by ojwb** [snowballstem/snowball#185]
I noticed pygment's Snowball support includes a test file which seems to be a Snowball stemmer for Esperanto:

https://github.com/pygments/pygments/blob/87bb672d4788186bc1d52faa8f5a7ccfb21c96ef/tests/examplefiles/snowball/example.sbl

I've not really investigated it but it seems plausible from an initial look.

@dscorbett It looks like you contributed the pygments support for Snowball (thanks for that) and your name is on the commit that added this test file - is this Esperanto stemmer yours?

If so, is it something we should be looking at merging?

One issue is it looks like the code has been deliberately tweaked to better serve as a test of the highlighting, e.g. `define short_word as not (loop (maxint * 0 + 4 / 2) gopast vowel)` and use of 3 different `stringescapes` settings.  Is there a "clean" version?  I didn't seem to turn up anything via searching the web.

**Comment by dscorbett** [comment]
Yes, I wrote that. I don’t know how much use an Esperanto stemmer would get, but it does work and it handles a lot of non-obvious cases. I designed it from the start to be a Pygments example file so there is no clean version. Should I clean it up and make PRs for here, snowball-data, and snowball-website?

**Comment by ojwb** [comment]
Sorry, I thought I'd responded at the time but it seems not.

> it does work and it handles a lot of non-obvious cases.

Given that, I think it's worth merging.

**Comment by dscorbett** [comment]
I’ll start cleaning this up soon. I haven’t checked but I think there is a bug with <i>naŭa</i>; I might have to implement an R1/R2 distinction to fix it. I should also double-check <i>ili</i>. I don’t know of any other bugs but I would like some feedback on some design decisions before I get back to this.

<i>-aŭ</i> is a weird suffix in Esperanto. It (mostly) can’t be separated from its root like other suffixes and therefore could be considered part of the root. On the other hand, the vocabulary is constructed such that removing it (usually) doesn’t produce another attested stem, so even if removing it could be considered overstemming, it’s not a real problem. (A counterexample is <i>palaŭa</i> ‘Palauan’ vs. <i>pala</i> ‘pale’.) My algorithm removes <i>-aŭ</i> because it was a fun technical challenge; I recommend keeping that behavior but I don’t feel strongly about it.

Derivational suffixes are generally ambiguous. For example, <i>arbaro</i> is derived from <i>arbo</i> but <i>elfaro</i> is unrelated to <i>elfo</i>. My algorithm avoids overstemming by skipping all derivational suffixes. That might cause some surprising understemmings but I think it is unavoidable.

Inflectional suffixes are almost entirely unambiguous. However, I am not sure about the participle suffixes. For example, <i>gravito</i> is unrelated to <i>gravi</i>, <i>dianto</i> is unrelated to <i>dio</i>, and <i>karoto</i> is unrelated to <i>kara</i>. I think such accidental ambiguities are uncommon but I don’t know. The algorithm currently ignores participle suffixes. What do you recommend?

The algorithm stems all parts of speech together because it is temptingly convenient, e.g. <i>morti</i> ‘to die’, <i>morto</i> ‘death’, <i>morta</i> ‘dead’, and <i>morte</i> ‘mortally’. Should some of these be kept separate?

**Comment by ojwb** [comment]
> I don’t know of any other bugs but I would like some feedback on some design decisions before I get back to this.
> 
> -aŭ is a weird suffix in Esperanto. It (mostly) can’t be separated from its root like other suffixes and therefore could be considered part of the root. On the other hand, the vocabulary is constructed such that removing it (usually) doesn’t produce another attested stem, so even if removing it could be considered overstemming, it’s not a real problem. (A counterexample is palaŭa ‘Palauan’ vs. pala ‘pale’.) My algorithm removes -aŭ because it was a fun technical challenge; I recommend keeping that behavior but I don’t feel strongly about it.

I guess the algorithm could perhaps be simpler if it exploited being mostly able to almost always remove `-aŭ`, but with an exception something like `not among ('pal' /*etc*/)` to exclude any problematic cases.  It seems cleaner to give a more correct linguistic stem though, unless doing so is a lot more complicated.

> Derivational suffixes are generally ambiguous. For example, arbaro is derived from arbo but elfaro is unrelated to elfo. My algorithm avoids overstemming by skipping all derivational suffixes. That might cause some surprising understemmings but I think it is unavoidable.

There are probably cases that are left alone because of ambiguities by most if not all the existing algorithms.  It's a bit unsatisfying, but it seems even constructed languages are messy!
 
> Inflectional suffixes are almost entirely unambiguous. However, I am not sure about the participle suffixes. For example, gravito is unrelated to gravi, dianto is unrelated to dio, and karoto is unrelated to kara. I think such accidental ambiguities are uncommon but I don’t know. The algorithm currently ignores participle suffixes. What do you recommend?

I don't know much about the structure of Esparanto's grammar, but I'd probably lean towards leaving them alone as unwanted conflation is much more problematic than missed conflation.  

Once we have a good vocabulary list we have a script (`scripts/stemmer-compare` in repo `snowball-data`) which can show the effect of a proposed change to a stemmer looking at the grouping of input words by their stems and how that changes, so we could evaluate the positive and negative effects of removing participle suffixes (or any other adjustment to the algorithm).

> The algorithm stems all parts of speech together because it is temptingly convenient, e.g. morti ‘to die’, morto ‘death’, morta ‘dead’, and morte ‘mortally’. Should some of these be kept separate?

Of those, 'mortally' perhaps seems something of an outlier.  Maybe that's down to nuances of meaning in the translations though.  At least based on the English meanings it doesn't seem terrible to conflate it, but it would be reasonable not to as well.

**Comment by ojwb** [comment]
#214 has been merged so closing.
